contracts/modules/royalty/policies/IpPool.sol

```diff
@@ -0,0 +1,225 @@
+
+    /// @notice Amount of tokens claimable at a given snapshot
+    mapping(uint256 snapshotId => mapping(address token => uint256 amount)) public claimableAtSnapshot;
+
+    /// @notice Amount of unclaimed tokens at the snapshot
+    mapping(uint256 snapshotId => uint32 tokenAmount) public unclaimedAtSnapshot;
+
+    /// @notice Indicates whether the claimer has claimed the revenue tokens at a given snapshot
+    mapping(uint256 snapshotId => mapping(address claimer => mapping(address token => bool)))
+        public isClaimedAtSnapshot;
+
+    /// @notice Royalty tokens of the pool
+    EnumerableSet.AddressSet private _tokens;
+
+    // TODO: change to beacon upgradeable contract
+    /// @notice The constructor of the IpPool
+    /// @param name The name of the pool token
+    /// @param symbol The symbol of the pool token
+    /// @param royaltyPolicyLAP The address of the royalty policy LAP
+    /// @param supply The total supply of the pool token
+    /// @param unclaimedTokens The amount of unclaimed tokens reserved for ancestors
+    /// @param ipId The ip id the pool belongs to
+    constructor(
+        string memory name,
+        string memory symbol,
+        address royaltyPolicyLAP,
+        uint32 supply,
+        uint32 unclaimedTokens,
+        address ipId
+    ) ERC20(name, symbol) {
+        if (ipId == address(0)) revert Errors.IpPool__ZeroIpId();
+        if (royaltyPolicyLAP == address(0)) revert Errors.IpPool__ZeroRoyaltyPolicyLAP();
+
+        ROYALTY_POLICY_LAP = IRoyaltyPolicyLAP(royaltyPolicyLAP);
+        IP_ID = ipId;
+        lastSnapshotTimestamp = block.timestamp;
+        unclaimedRoyaltyTokens = unclaimedTokens;
+
+        _mint(address(this), unclaimedTokens);
+        _mint(ipId, supply - unclaimedTokens);
+    }
+
+    /// @notice Adds a new revenue token to the pool
+    /// @param token The address of the revenue token
+    /// @dev Only callable by the royalty policy LAP
+    function updateIpPoolTokens(address token) external {
+        if (msg.sender != address(ROYALTY_POLICY_LAP)) revert Errors.IpPool__NotRoyaltyPolicyLAP();
+        _tokens.add(token);
+    }
+
+    /// @notice Snapshots the claimable revenue and royalty token amounts
+    /// @return snapshotId The snapshot id
+    function snapshot() external returns (uint256) {
+        if (block.timestamp - lastSnapshotTimestamp < ROYALTY_POLICY_LAP.getSnapshotInterval())
+            revert Errors.IpPool__SnapshotIntervalTooShort();
+
+        uint256 snapshotId = _snapshot();
+        lastSnapshotTimestamp = block.timestamp;
+
+        uint32 unclaimedTokens = unclaimedRoyaltyTokens;
+        unclaimedAtSnapshot[snapshotId] = unclaimedTokens;
+
+        address[] memory tokens = _tokens.values();
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 tokenBalance = IERC20(tokens[i]).balanceOf(address(this));
+            if (tokenBalance == 0) {
+                _tokens.remove(tokens[i]);
```

Comment:
Do we want to remove a token from the set when its balance is zero? What if someone sends it right after taking a snapshot (that removes the token from the set)

Reply:
It is added again when such payment is made via updateIpPoolTokens which is called in payRoyaltyOnBehalf. The idea is to keep only the necessary list of tokens to avoid looping unnecessarily.